### PR TITLE
DOC: Explain how to use multiple adapters at the same time

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Explain how to use multiple adapters (e.g. 2 LoRA adapters) at the same time, as the API is not quite intuitive and there are some footguns around trainable parameters.

This question has come up multiple times in the past (for recent examples, check #2749 and #2756). Thus it's a good idea to properly document this.

### CODE REVIEW

File: docs/source/developer_guides/troubleshooting.md [hunk 1]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
 Assume the error message says that the unknown keyword argument is named `foobar`. Search inside the `adapter_config.json` of this PEFT adapter for the `foobar` entry and delete it from the file. Then save the file and try loading the model again.
 
 This solution works most of the time. As long as it is the default value for `foobar`, it can be ignored. However, when it is set to some other value, you will get incorrect results. Upgrading PEFT is the recommended solution.
+
+## Adapter handling
+
+### Using multiple adapters at the same time
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md:407]:
```suggestion
## Using multiple adapters at the same time
```

File: docs/source/developer_guides/troubleshooting.md [hunk 2]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
...
+
+PEFT allows you to create more than one adapter on the same model. This can be useful in many situations. For example, for inference, you may want to serve two fine-tunes using the same base model instead of loading the base model once for each fine-tune, which would cost more memory. However, a less well known feature is that multiple adapters can be activated at the same time. This way, the model may leverage the learnings from all those adapters at the same time. As an example, if you have a diffusion model, you may want to use one LoRA adapter to change the style and a different one to change the subject.
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md]:
```suggestion
PEFT allows you to create more than one adapter on the same model. This can be useful in many situations. For example, for inference, you may want to serve two fine-tuned models from the same base model instead of loading the base model once for each fine-tuned model, which would cost more memory. However, multiple adapters can be activated at the same time. This way, the model may leverage the learnings from all those adapters at the same time. As an example, if you have a diffusion model, you may want to use one LoRA adapter to change the style and a different one to change the subject.
```

File: docs/source/developer_guides/troubleshooting.md [hunk 3]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
...
+
+Activating multiple adapters at the same time is generally possible on all PEFT methods except for prompt learning methds (p-tuning, prefix tuning, etc.). Therefore, it works for LoRA, LoHa, IA³ etc. The following example illustrates how to achieve this:
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md]:
```suggestion
Activating multiple adapters at the same time is generally possible on all PEFT methods (LoRA, LoHa, IA³, etc.) except for prompt learning methods (p-tuning, prefix tuning, etc.). The following example illustrates how to achieve this:
```

File: docs/source/developer_guides/troubleshooting.md [hunk 4]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
...
+model.base_model.set_adapter(["default", "other"])
+```
+
+<Tip>
+
+In the example above, you can see that we need to call `model.base_model.set_adapter(["default", "other"])`. Why can we not call `model.set_adapter(["default", "other"])`? This is unfortunately not possible because, as explained earlier, some PEFT methods don't support activating more than one adapter at a time.
+
+</Tip>
+
+Some of you may also want to train two adapters at the same time, e.g. training one LoRA adapter on one loss and the other one on another loss. This is also possible but you should be careful to ensure that the weights of both adapters are known to the optimizer. Otherwise, only one adapter will receive updates.
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md]:
```suggestion
It is also possible to train two adapters at the same time, but you should be careful to ensure that the weights of both adapters are known to the optimizer. Otherwise, only one adapter will receive updates.
```

File: docs/source/developer_guides/troubleshooting.md [hunk 5]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
...
+trainer.train()
+```
+
+or
+
+```python
+optimizer = torch.optim.AdamW([param for param in model.parameters() if param.requires_grad], ...)
+```
+
+then the second LoRA adapter (`"other"`) would not be trained. This is because it is inactive at this moment, thus the `requires_grad` attribute on its parameters is set to `False` and the optimizer will ignore it. Therefore, ensure to activate all adapters that should be trained _before_ initializing the optimizer:
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md]:
```suggestion
then the second LoRA adapter (`"other"`) would not be trained. This is because it is inactive at this moment, which means the `requires_grad` attribute on its parameters is set to `False` and the optimizer will ignore it. Therefore, make sure to activate all adapters that should be trained _before_ initializing the optimizer:
```

File: docs/source/developer_guides/troubleshooting.md [hunk 6]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
...
+
+model_id = ...
+base_model = AutoModelForCausalLM.from_pretrained(model_id)
+model = PeftModel-from_pretrained(base_model, lora_path_0)  # default adapter_name is 'default'
+model.load_adapter(lora_path_1, adapter_name="other")
+# the 'other' adapter was loaded but it's not active yet, so to activate both adapters:
+model.base_model.set_adapter(["default", "other"])
+```
+
+<Tip>
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md]: Minor nit, but it may be nice to reduce the number of lines to use this instead of `<Tip> ... </Tip>`

```md
> [!TIP]
> ...
```

File: docs/source/developer_guides/troubleshooting.md [hunk 7]
```diff
@@ -401,3 +401,73 @@ If it is not possible for you to upgrade PEFT, there is a workaround you can try
...
+
+Activating multiple adapters at the same time is generally possible on all PEFT methods except for prompt learning methds (p-tuning, prefix tuning, etc.). Therefore, it works for LoRA, LoHa, IA³ etc. The following example illustrates how to achieve this:
+
+```python
+from transformers import AutoModelForCausalLM
+from peft import PeftModel
+
+model_id = ...
+base_model = AutoModelForCausalLM.from_pretrained(model_id)
+model = PeftModel-from_pretrained(base_model, lora_path_0)  # default adapter_name is 'default'
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md]:
```suggestion
model = PeftModel.from_pretrained(base_model, lora_path_0)  # default adapter_name is 'default'
```

**Review by stevhliu — APPROVED** [review]
Thanks for adding new docs!

File: docs/source/developer_guides/troubleshooting.md (hunk 1 above)
Comment (BenjaminBossan) [reply docs/source/developer_guides/troubleshooting.md:407]: It's supposed to be a subsection of "Adapter handling", so `###`.

File: docs/source/developer_guides/troubleshooting.md (hunk 6 above)
Comment (BenjaminBossan) [reply docs/source/developer_guides/troubleshooting.md]: Changed.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the review @stevhliu
